### PR TITLE
Added parser for SPDX files 

### PR DESCRIPTION
**Pull request by aasim**

### CODE REVIEW

File: test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/TestParser.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+using Microsoft.Sbom.Parsers.Spdx22SbomParser;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+
+namespace Microsoft.Sbom.Parser
+{
+    internal class TestParser
```
Comment (aasim) [inline test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/TestParser.cs:11]: This is a temporary testing class, once the final parser is written, I will update this to use that class instead.

File: test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/TestParser.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        public IEnumerable<SBOMFile> GetFiles(Stream stream)
+        {        
+            stream.Read(buffer);
+
+            while (GetFiles(stream, out SBOMFile sbomFile) != 0)
+            {
+                yield return sbomFile;
+            }
+
+            long GetFiles(Stream stream, out SBOMFile sbomFile)
```
Comment (aasim) [inline test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/TestParser.cs:26]: A ref struct can't be used inside an iterator method, so used an inline method that has its own stack.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Constants.cs [hunk 3]
```diff
@@ -61,5 +61,28 @@ internal class Constants
                                                             Name = Constants.SPDXName,
                                                             Version = Constants.SPDXVersion
                                                         };
+
+        internal const int ReadBufferSize = 30;
+        internal const int MaxReadBufferSize = 32768;
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Constants.cs:66]: Why these values?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    /// <summary>
+    /// Read the next JSON token in the reader from the input buffer.
+    /// If the buffer is small and doesn't contain all the text for the next token, 
+    /// a call to GetMoreBytesFromStream is made to read more data into the buffer.
+    /// </summary>
+    /// <exception cref="EndOfStreamException"></exception>
+    public static void Read(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+    {
+        // If the buffer is empty, refill the buffer.
+        if (buffer[0] == 0)
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:27]: >buffer[0] [](http://example.com/codeflow?start=12&length=9)

Could this throw if buffer is empty?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+/// </summary>
+internal class ParserUtils
+{
+    /// <summary>
+    /// Read the next JSON token in the reader from the input buffer.
+    /// If the buffer is small and doesn't contain all the text for the next token, 
+    /// a call to GetMoreBytesFromStream is made to read more data into the buffer.
+    /// </summary>
+    /// <exception cref="EndOfStreamException"></exception>
+    public static void Read(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:24]: >stream [](http://example.com/codeflow?start=35&length=6)

check for null

Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:24]: >Utf8JsonReader [](http://example.com/codeflow?start=66&length=14)

Is SPDX UTF8?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    /// <summary>
+    /// Asserts if the reader is at the current expected token.
+    /// </summary>
+    /// <exception cref="ParserError"></exception>
+    internal static void AssertTokenType(Stream stream, ref Utf8JsonReader reader, JsonTokenType expectedTokenType)
+    {
+        if (reader.TokenType != expectedTokenType)
+        {
+            throw new ParserError($"Expected a '{Constants.JsonTokenStrings[(byte)expectedTokenType]}' at position {stream.Position}");
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:50]: >(byte)expectedTokenType [](http://example.com/codeflow?start=76&length=23)

How can you guarantee that Constants.JsonTokenStrings contains the element you are looking for? What is the relationship between JsonTokenType and Constants.JsonTokenStrings?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        }
+        catch (EndOfStreamException)
+        {
+            sbomFile = null;
+            return 0;
+        }
+        catch (JsonException e)
+        {
+            sbomFile = null;
+            throw new ParserError($"Error while parsing JSON, addtional details: ${e.Message}");
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:79]: >ParserError [](http://example.com/codeflow?start=22&length=11)

Maybe add the JsonException and InnerException?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Constants.cs (hunk 3 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Constants.cs:66]: Oh good catch, these were for testing, I changed it back to 4096.

This size should be greater than the largest token you can have in your JSON file. 4096 is a good buffer size in general and should accommodate all the parameters that we have

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        {
+            throw new ParserError($"Missing required value(s) for file object at position {stream.Position}: {string.Join(",", missingProps)}");
+        }
+    }
+
+    private void ParseProperty(ref Utf8JsonReader reader, ref byte[] buffer)
+    {
+        switch (reader.GetString())
+        {
+            case "fileName":
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:129]: >"fileName" [](http://example.com/codeflow?start=17&length=10)

Can you use consts?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs (hunk 4 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:37]: Yes, fixed by adding null check

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs (hunk 6 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:65]: Its an enum, and I have enumerated every element in the enum, so it should always have the required element

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        }
+    }
+
+    private void ValidateSbomFile(SBOMFile sbomFile)
+    {
+        // I want to use the DataAnnotations Validator here, but will check with CB first
+        // before adding a new dependency.
+        var missingProps = new List<string>();
+
+        if (sbomFile.Checksum == null || sbomFile.Checksum.Where(c => c.Algorithm.Name == "SHA256").Count() == 0)
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:94]: >"SHA256" [](http://example.com/codeflow?start=90&length=8)

Don't we have an enum for this?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+        while (reader.TokenType != JsonTokenType.EndArray)
+        {
+            ParserUtils.Read(stream, ref buffer, ref reader);
+            if (reader.TokenType == JsonTokenType.EndArray)
+            {
+                break;
+            }
+
+            licenses.Add(reader.GetString());
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:185]: You could also have the following code to avoid the break:
if (reader.TokenType == JsonTokenType.EndArray)
{
       licenses.Add(reader.GetString());
}

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    private List<string> ParseLicenseInfoInFilesArray(ref Utf8JsonReader reader, ref byte[] buffer)
+    {
+        var licenses = new List<string>();
+
+        // Read the opening [ of the array
+        ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.StartArray);
+
+        while (reader.TokenType != JsonTokenType.EndArray)
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:177]: Would this loop indefinetely if there is no EndArray token?

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+
+        // Move to the first property token
+        ParserUtils.Read(stream, ref buffer, ref reader);
+        ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.PropertyName);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            switch (reader.GetString())
+            {
+                case "algorithm":
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:230]: >"algorithm" [](http://example.com/codeflow?start=21&length=11)

Enums or consts?

File: test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/Strings/JsonStrings.cs [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+                ],
+                ""licenseConcluded"": ""NOASSERTION"",
+                ""licenseInfoInFiles"": [
+                    ""NOASSERTION"", ""GNU""
+                ],
+                ""copyrightText"": ""NOASSERTION""
+            }
+        ]";
+
+    public const string MalformedJson = @"
```
Comment (ByAgenT) [inline test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/Strings/JsonStrings.cs:50]: Maybe also add more negative tests? Empty JSON (`{}`,`[]`), empty content these two comes to my mind

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs (hunk 5 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:24]: Fixed

Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/ParserUtils.cs:24]: Yes, its required to be UTF8

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 7 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:84]: Fixed

File: src/Microsoft.Sbom.Contracts/Contracts/SBOMFile.cs [hunk 14]
```diff
@@ -34,5 +34,10 @@ public class SBOMFile
         /// Gets or sets the relative path to the BuildDropPath of the file in the SBOM.
         /// </summary>
         public string Path { get; set; }
+
+        /// <summary>
+        /// Gets or sets unique Identifier for the file.
+        /// </summary>
+        public string SPDXId { get; set; }
```
[the remark this answers is not in the document]
Comment (aasim) [reply src/Microsoft.Sbom.Contracts/Contracts/SBOMFile.cs:41]: Right, I had remembered to change it before sending out the PR... yes this should be just Id here.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 11 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:177]: No, the reader call will fail once we reach end of stream, I will add a test

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+
+    public SbomFileParser(Stream stream)
+    {
+        this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
+
+        sbomFile = new ();
+    }
+
+    /// <summary>
+    /// Returns an iterator of <see cref="SBOMFile"/> objects by parsing the SPDX JSON files section.
```
[the remark this answers is not in the document]
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:31]: Fixed

File: test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/Strings/JsonStrings.cs (hunk 13 above)
Comment (aasim) [reply test/Microsoft.Sbom.Parsers.Spdx22SbomParser.Tests/Parser/Strings/JsonStrings.cs:50]: Fixed

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 10 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:185]: The licenses.Add method and the line before this == both read to the buffer which is a mutable call and I added the break just to avoid this.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 8 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:134]: Fixed

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 9 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs]: Fixed

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+        }
+
+        if (sbomFile.LicenseInfoInFiles == null || sbomFile.LicenseInfoInFiles.Count == 0)
+        {
+            missingProps.Add(nameof(sbomFile.LicenseInfoInFiles));
+        }
+
+        if (missingProps.Count() > 0)
+        {
+            throw new ParserError($"Missing required value(s) for file object at position {stream.Position}: {string.Join(",", missingProps)}");
```
Comment (ksigmund) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:141]: I personally tend to prefer returning something like `ParserResult` and avoid throwing until the last possible moment.  It makes testing somewhat easier and given that all of these scenarios are expected, it's hard to classify them as truly "exceptional."  Just a thought.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Exceptions/ParserError.cs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+using System;
+using System.Runtime.Serialization;
+
+namespace Microsoft.Sbom.Exceptions;
+
+/// <summary>
+/// Thrown when the SPDX parser detects an error in the JSON file.
+/// </summary>
+[Serializable]
+public class ParserError : Exception
```
Comment (ksigmund) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Exceptions/ParserError.cs:13]: I think exceptions generally are named something like i.e. ParserException

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs (hunk 16 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:141]: hmm.. I like that as well, might be faster as well, since throwing and catching is expensive. Did you think of something like a Monad IResult type?

Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs:141]: lets merge this PR as is, and I will send out a new PR with the results type in it. That will simplify some other stuff as well.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Exceptions/ParserError.cs (hunk 17 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Exceptions/ParserError.cs]: Fixed

**Review by ByAgenT — APPROVED** [review]
